tensorflow/stream_executor/matmul_util.h

```diff
@@ -0,0 +1,255 @@
+                           dtype_cd_, allow_tf32_, device_id_);
+  }
+
+  bool trans_a_;
+  bool trans_b_;
+  bool adj_a_;
+  bool adj_b_;
+  uint64 m_;
+  uint64 n_;
+  uint64 k_;
+  uint64 batch_count_;
+  bool broadcast_a_;
+  bool broadcast_b_;
+  tensorflow::DataType dtype_ab_;
+  tensorflow::DataType dtype_cd_;
+  bool allow_tf32_;
+  int device_id_;
+  uint64 hash_code_;
+};
+
+static inline port::StatusOr<blas::ComputationType> GetBlasComputationType(
+    const tensorflow::DataType& dtype, bool allow_tf32) {
+  using blas::ComputationType;
+  static bool use_f32_for_f16_computation =
+      tensorflow::MatmulDoFP32ComputationFP16Input();
+  ComputationType f32_type =
+      allow_tf32 ? ComputationType::kTF32AsF32 : ComputationType::kF32;
+  switch (dtype) {
+    case tensorflow::DT_HALF:
+    case tensorflow::DT_BFLOAT16:
+      return use_f32_for_f16_computation ? f32_type : ComputationType::kF16;
+    case tensorflow::DT_FLOAT:
+      return f32_type;
+    case tensorflow::DT_DOUBLE:
+      return ComputationType::kF64;
+      return port::Status::OK();
+    case tensorflow::DT_COMPLEX64:
+      return f32_type;
+    case tensorflow::DT_COMPLEX128:
+      return ComputationType::kComplexF64;
+    default:
+      // Unsupported compute_type, return false.
+      return port::InternalError("Unsupported dtype for batched matmul 2");
+  }
+}
+
+// Thread-safe map from matmul parameters to their corresponding plan and
+// algorithms.
+template <typename Parameters>
+class BlasLtMatmulPlanMap {
+ public:
+  const blas::PlanAndAlgorithms* Find(const Parameters& params) {
+    tensorflow::mutex_lock lock(mu_);
+    auto iter = params_plan_map_.find(params);
+    if (iter == params_plan_map_.end()) {
+      return nullptr;
+    }
+    return &iter->second;
+  }
+  const blas::PlanAndAlgorithms* Insert(const Parameters& params,
+                                        blas::PlanAndAlgorithms value) {
+    tensorflow::mutex_lock lock(mu_);
+    return &params_plan_map_.emplace(params, std::move(value)).first->second;
+  }
+
+ private:
+  struct Hasher {
+    std::size_t operator()(const Parameters& parameter) const {
+      return parameter.hash();
+    }
+  };
+
+  mutable tensorflow::mutex mu_;
+  std::unordered_map<Parameters, blas::PlanAndAlgorithms, Hasher>
+      params_plan_map_ GUARDED_BY(mu_);
+};
+
+template <typename Parameters>
+struct BlasLtPlanMapSingleton {
+  typedef BlasLtMatmulPlanMap<Parameters> PlanMapType;
+  static PlanMapType* GetInstance() {
+    static PlanMapType* instance = new PlanMapType();
+    return instance;
+  }
+};
+
+typedef BlasLtPlanMapSingleton<BatchMatmulParameters>
+    BatchMatmulPlanMapSingleton;
+
+// Dummy type to group matmul autotune results together.
+struct BatchMatmulAutoTuneGroup {
+  static string name() { return "MatmulLt"; }
+};
+
+typedef tensorflow::AutotuneSingleton<
+    BatchMatmulAutoTuneGroup, BatchMatmulParameters, blas::AlgorithmConfig>
+    AutoTuneBatchMatmul;
```

Comment:
All other autotuning maps from XLA don't use this helper class from TF, using it just for cublasLT would be inconsistent. Any particular features from that class you would like to use ?
Also @awpr any thoughts?